Add native support for wxOverlay under wxGTK3

AliKet opened the pull request:
This is a **draft** and **WIP**

I made this PR for two reasons:

1. to share my idea of how to implement `wxOverlay` under `wxGTK` (GTK3 in fact)
2. to ask questions and get some help to implement this correctly.

I tested this PR with these : **grid**, **caret** and **drawing** samples
under **Ubuntu, Kubuntu and Mint** 18.04. and under **Weston** (a `wayland` compositor) and
they seem to be working fine.

TIA.

AliKet commented:
@vadz , @paulcor 

Can you take a look at this and tell me your thoughts about it ?

TIA.

AliKet commented:
The caret sample is not working correctly because the old position/rectangle is not refreshed/invalidated at all.

I managed to solve this problem in two ways:

1. If the new rectangle passed to wxOverlay (e.g. after a Reset()) has changed, wxOverlay will refresh/invalidate both rectangles, the new and the old one.
2. Blink out the caret before moving the it to its new position.

The two options above worked well and i don't know which one to choose  ?

vadz commented on include/wx/generic/grid.h:
```diff
@@ -16,13 +16,19 @@
     #include <iterator>
 #endif
 
+#if defined(__WXGTK3__)
```
I'd like to test this under Mac to see if it also works with native overlay implementation there. It really ought to, and then we wouldn't need to have this GTK-specific test and could just use `wxHAS_NATIVE_OVERLAY`.

Although, OTOH, I al;so wonder why shouldn't we always use overlay, i.e. why shouldn't it work even with the generic implementation?

AliKet replied:
I'm focusing on getting wxOverlay working decently under GTK3, i'll test with the generic version later.

vadz commented on include/wx/overlay.h:
```diff
@@ -17,6 +17,8 @@
     #define wxHAS_NATIVE_OVERLAY 1
 #elif defined(__WXOSX__) && wxOSX_USE_COCOA
     #define wxHAS_NATIVE_OVERLAY 1
+#elif defined(__WXGTK3__) // We should test for GTK_CHECK_VERSION(3,18,0)
```
This needs to be done at configure level, similarly to what we do for e.g. `__WXGTK218__`, i.e. we need to define `__WXGTK318__` there which would be available in this header (unlike GTK version-checking macros).

AliKet replied:
Can you show me how to do that, please ?

AliKet replied:
Added in this commit e665298.

vadz commented on src/gtk/window.cpp:
```diff
@@ -2702,20 +2702,63 @@ bool wxWindowGTK::Create( wxWindow *parent,
+        gtk_container_add( GTK_CONTAINER(overlay), m_wxwindow );
+        gtk_container_add( GTK_CONTAINER(m_widget), overlay );
+    }
+    else // m_widget == m_wxwindow
```
Does this really happen for non standard classes?

vadz commented on src/gtk/window.cpp:
```diff
@@ -2702,20 +2702,63 @@ bool wxWindowGTK::Create( wxWindow *parent,
+        m_widget = overlay;
+    }
+
+    gtk_widget_show_all(m_widget);
```
We didn't do this before, her,e but only in `PostCreation()`. Does it really need to be done here now? If so, it should at least check `m_isShown`, as the existing call does.

vadz commented on include/wx/gtk/window.h:
```diff
@@ -207,6 +207,8 @@ class WXDLLIMPEXP_CORE wxWindowGTK : public wxWindowBase
     virtual void GTKHandleRealized();
     void GTKHandleUnrealize();
 
+    void GTKEnableOverlay();
```
Should have a comment explaining that it does nothing if overlay can't be enabled. Maybe even rename it to `GTKEnableOverlayIfPossible()`. OTOH it would be even better to avoid calling it for all windows as discussed above.

AliKet replied:
I changed it's implementation and now it returns `false` on failure.

vadz commented on include/wx/gtk/private/overlay.h:
```diff
@@ -0,0 +1,58 @@
+
+#include "wx/gdicmn.h"
+
+typedef struct _cairo cairo_t;
+typedef struct _cairo_surface cairo_surface_t;
```
FWIW you can include Cairo headers directly from _private_ headers.

vadz commented on include/wx/gtk/private/overlay.h:
```diff
@@ -0,0 +1,58 @@
+
+    cairo_t*         m_cairoContext;
+    cairo_surface_t* m_cairoSurface;
+};
```
Should probably have
```suggestion
    wxDECLARE_NO_COPY_CLASS(wxOverlayImpl);
};
```
here.

vadz commented on src/gtk/overlay.cpp:
```diff
@@ -0,0 +1,222 @@
+    wxCHECK_RET( dc, "NULL dc pointer" );
+    wxASSERT_MSG( !IsOk() , "You cannot Init an overlay twice" );
+
+    wxLogDebug("wxOverlayImpl::Init(%d, %d, %d, %d)", x, y, width, height);
```
Please remove this in the final version or replace it with `wxLogTrace("overlay", ...)` if you think it can be useful to keep it.

vadz commented on src/gtk/overlay.cpp:
```diff
@@ -0,0 +1,222 @@
+    gtk_widget_set_can_focus(m_surface, false);
+
+    gtk_overlay_add_overlay( GTK_OVERLAY(overlay), m_surface );
+    gtk_overlay_set_overlay_pass_through( GTK_OVERLAY(overlay), m_surface, true );
```
Minor, but
```suggestion
    gtk_overlay_set_overlay_pass_through( GTK_OVERLAY(overlay), m_surface, TRUE );
```

vadz commented on src/gtk/overlay.cpp:
```diff
@@ -0,0 +1,222 @@
+void wxOverlayImpl::OnSize(wxSizeEvent& event)
+{
+    Reset();
+    event.Skip(); // wxCaret sample crashes without this!
```
Comment is a bit superfluous, you're supposed to skip size events unless you explicitly want to prevent the default handling from taking place, which is not the case here.

vadz commented on src/gtk/overlay.cpp:
```diff
@@ -0,0 +1,222 @@
+    // Create the surface:
+
+    GtkWidget* const overlay = GetOverlayWidget(m_window);
+    wxASSERT_MSG( overlay, "Overlay is not enabled for m_window" );
```
Should be a `CHECK` to avoid crashing immediately below.

vadz reviewed (changes requested):
Thanks a lot for working on this, it would be definitely great to have this in 3.2.0!

The main problem I have is with `WXUSE_OVERLAY`. I'd really like to get rid of it (if we do need to keep it, we must document it in `docs/doxygen/overviews/envvars.h` and mention it in `wxOverlay` documentation so that people could use it in their own class).

Ideal would be to create the overlay on demand, i.e. the first time `wxDCOverlay` is used with the paint DC associated with the given window. If we can't do this, we need to add a flag and/or member function enabling the use of the overlay and complain if this flag is not set when `wxDCOverlay` is created, but this will come with the usual problem of breaking (by showing asserts for it) the existing code which used to work fine (and still would work fine under MSW/Mac), so the automatic solution would be really preferable. So what exactly prevents is from creating the overlay later, and reparenting the widget under it?

Thanks again!

paulcor commented:
Having never really looked at `wxOverlay` before, I have spent what seems like an inordinate amount of time trying to understand how it is supposed to work, as the documentation is almost no help at all. Let's see if I've got this right:

You need an `wxOverlay` lying around for your window, but you don't directly associate the two. That happens through the magic of `wxDCOverlay`. Which is not a `wxDC`. Instead, it hijacks an existing `wxDC` (there's the implicit connection to your window) so that subsequent drawing with that dc will go to the overlay layer. That layer then appears on top of your window. Oh, and BTW, if I'm understanding the wxOSX and wxDFB implementations correctly, when `wxDCOverlay` is done with it, the dc is broken and can no longer be used to draw on the window.

In current usage in the wx sources, the dc is always a `wxClientDC` or `wxScreenDC`, created explicitly and exclusively for `wxDCOverlay`.

paulcor commented on src/gtk/overlay.cpp:
```diff
@@ -0,0 +1,222 @@
+        cairo_pop_group_to_source(m_cairoContext);
+        cairo_surface_set_device_offset(m_cairoSurface, 0, 0);
+
+        m_window->RefreshRect(m_rect);
```
If the window is going to be re-painted, I don't understand the point of using `GtkOverlay`. Why not just draw an overlay surface at the end of the paint?

AliKet commented on src/gtk/overlay.cpp:
```diff
@@ -0,0 +1,200 @@
+#ifdef wxHAS_NATIVE_OVERLAY
+
+#define wxDEBUG_GTK_OVERLAY 0 // set this to 1 to see the overlay
+
```
Leftover

AliKet commented:
> Having never really looked at `wxOverlay` before, I have spent what seems like an inordinate amount of time trying to understand how it is supposed to work, as the documentation is almost no help at all. Let's see if I've got this right:
> 

TBH, I didn't know at first how the **GTK** version could be implemented following the current API.
but fortunately the `wxDFB` implementation came to the rescue. from which i inspired this implementation.

> You need an `wxOverlay` lying around for your window, but you don't directly associate the two. That happens through the magic of `wxDCOverlay`. Which is not a `wxDC`. Instead, it hijacks an existing `wxDC` (there's the implicit connection to your window) so that subsequent drawing with that dc will go to the overlay layer. That layer then appears on top of your window. Oh, and BTW, if I'm understanding the wxOSX and wxDFB implementations correctly, when `wxDCOverlay` is done with it, the dc is broken and can no longer be used to draw on the window.
> 
Exactly, with the gtk version the hijacking happens by temporarily redirects drawing to an intermediate surface which will be drawn to the created layer (GtkDrawingArea).
but with the help of cairo_save() and cairo_restore() i think the source DC should be intact after wxOverlay has done with it (to be demonstrated).

> In current usage in the wx sources, the dc is always a `wxClientDC` or `wxScreenDC`, created explicitly and exclusively for `wxDCOverlay`.

AliKet commented on src/gtk/overlay.cpp:
```diff
@@ -0,0 +1,210 @@
+    gtk_widget_show(m_surface);
+
+    m_window->SetFocus();
+}
```
@paulcor , @vadz 

Adding overlay to `m_window` is not supposed to alter the focus handling of that window in any way. but what i noticed is that:
- the first time i click on such window (`wxGrid` in fact, which supposed to set the focus on it) is not working at all!
- subsequent clicks work correctly.

I tried to trace the focus handling (with and without the `wxOverlay` being applied) but i'm still clueless what's going on here. any help/hint is much appreciated here of course.

For now, i'm adding an explicit call to `SetFocus()` which solves the problem. and if i am not mistaken, this can be an acceptable workaround knowing that `wxOverlay` expects `m_window` to be the focus window anyhow ?

paulcor commented:
Testing rubber-banding with the `drawing` sample shows that this PR currently causes full repaints of the window whenever the overlay layer changes, so it would still be no better than a much simpler "redraw-the-window" technique.

You can easily see the updates with X11 using `GTK_DEBUG=updates`.  That apparently doesn't work with Wayland, so you have to log paint events to get information on updates there.

AliKet commented:
> Testing rubber-banding with the `drawing` sample shows that this PR currently causes full repaints of the window whenever the overlay layer changes, so it would still be no better than a much simpler "redraw-the-window" technique.
> 

Thank you so much for the test, much appreciated.

> You can easily see the updates with X11 using `GTK_DEBUG=updates`. That apparently doesn't work with Wayland, so you have to log paint events to get information on updates there.

As rubber-banding needs a live updated rectangle to work with, we need to add a function `SetRectangle()` (or `UpdateRectangle()` or whatever this has to be named) and use it to update the rectangle. I'll push these changes shortly.

I think `wxGrid` could be improved this way too. but I'll leave this for now...

Thanks in advance for your collaboration on this.

paulcor commented:
IIUC, all you'll be able to do is restrict the window redraw to a given area, still no better than just redrawing that part of the window in a simpler way.

AliKet commented:
> IIUC, all you'll be able to do is restrict the window redraw to a given area, still no better than just redrawing that part of the window in a simpler way.

Simplicity, yes, this is why i patched my wxClientDC under wayland to do just that, i.e. issue a redraw in the dtor.
but FWIW here is one reason why i still think wxOverlay (for GTK) is worth the cost to be included into the library:
1- Abstruction, at least one have to write once and compile/run everywhere. and wxOverlay gives us this abstruction. Isn't that what wxWidgets is supposed to offer to its users?

and if efficiency is a concern for you, i already did a profiling to the wxGrid simple and wxOverlay is never a concern.

Anyway, i updated this PR if you want to try it out. TIA.

Regards.

vadz commented:
For me the important question is how does this compare to using the generic implementation. I think the native version should be more efficient because it hopefully doesn't allocate a bitmap containing the full window contents, but I'm not completely sure it doesn't do this, actually.

Another important question is whether the native version has any drawbacks compared to the generic one. Are you already aware of any?

And, on a somewhat related note, does the native version in the scenario of [this ticket](https://trac.wxwidgets.org/ticket/14415)?

paulcor commented:
> For me the important question is how does this compare to using the generic implementation.

I would say worse for X11, better for Wayland in that it works, as currently it does not.

> I think the native version should be more efficient because it hopefully doesn't allocate a bitmap containing the full window contents

Copying and drawing a bitmap is pretty quick and IMO likely more efficient than repainting a window, which can potentially be relatively expensive.

I'll note again that there is a simpler way to achieve the same effect as this PR, using basically the same technique `wxDFB` uses. Although I'd hardly call it "native", as it just repaints the window whenever the overlay changes. I'm looking into it.

AliKet commented:
> For me the important question is how does this compare to using the generic implementation. I think the native version should be more efficient because it hopefully doesn't allocate a bitmap containing the full window contents, but I'm not completely sure it doesn't do this, actually.
> 

Well, before i can give you a definitive answer about how this performs compared to the generic version i need to do some real profiling side by side and post the result here, hopefully.

but yeah, there is no bitmap allocated to save and restore. and what would be drawn to the DC is eventually painted to the overlay. 

> Another important question is whether the native version has any drawbacks compared to the generic one. Are you already aware of any?
> 

the only drawback i'm facing right now is the need to call `GTKEnableOverlay()` at the time the window wanting to use `wxOverlay` is constructed. in fact, i implemented `wxOverlay` so that the overlay widget is created on demand. but it doesn't work, unfortunately. i.e. the first interaction with the window doesn't work correctly:
- the window doesn't get focus from the first click.
- the window doesn't capture the mouse correctly.

even though subsequent interactions with the window works as expected.
I still don't know what's the problem here ? maybe just document this as limitation ?

> And, on a somewhat related note, does the native version in the scenario of [this ticket](https://trac.wxwidgets.org/ticket/14415)?

No, the rubber-banding is working just fine with this PR.

Best.

AliKet commented:
@paulcor , @vadz 

I left out using `wxOverlay` with `wxGrid` in favor of using something simpler (which i am not sure what it is right now) and changed `wxSplitterWindow` to use it for a non live sash update.

I noticed that in the **wxOSX** and **wxDFB** versions, `wxOverlay::Reset()` do a hard reset which is not quite optimal at best if the overlay is used frequently. so i added a parameter to this function to let the code using it decide if the overlay should be disposed or just hidden to be reused on the next run.
even the **wxOSX** version has this comment:

> // todo : don't dispose, only hide and reposition on next run

I discovered that `size_allocate` omits the case when the window is scrollable and just override the client size calculated by `wxWindowGTK::DoGetClientSize()` and this makes the auto-scrolling feature not working correctly in presence of the scrollbars.

Regard.

Commit:
Add native support for wxOverlay under wxGTK3

AliKet commented:
replaced by  #2522
